spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -174,41 +187,81 @@ protected Object build(Object payload, MessageHeaders headers) {
 
 		private final MimeType mimeType;
 
-		private ContentTypeConvertingInterceptor(String contentType) {
+		private final boolean input;
+
+		private final Class<?> klazz;
+
+		private final MessageWrappingMessageConverter messageConverter;
+
+		private ContentTypeConvertingInterceptor(String contentType, boolean input) {
 			this.contentType = contentType;
 			this.mimeType = MessageConverterUtils.getMimeType(contentType);
+			this.input = input;
+			if (MessageConverterUtils.X_JAVA_OBJECT.includes(this.mimeType)) {
+				this.klazz =
+						MessageConverterUtils
+								.getJavaTypeForJavaObjectContentType(this.mimeType);
+			}
+			else if (this.mimeType.equals(MessageConverterUtils.X_SPRING_TUPLE)) {
+				this.klazz = Tuple.class;
+			}
+			else if (this.mimeType.getType().equals("text") || this.mimeType.getSubtype().equals(
+					"json") || this.mimeType.getSubtype().equals("xml")) {
+				this.klazz = String.class;
+			}
+			else {
+				this.klazz = byte[].class;
+			}
+			if (input) {
```

Comment:
Yes, I thought that the two would be different, but at this point they don't seem to.